cmd/shipper/main.go

```diff
@@ -1,16 +1,106 @@
 package main
 
 import (
+	"gz-beat-shipper/configs"
+	"gz-beat-shipper/pkg/filehandler"
+	"gz-beat-shipper/pkg/registry"
 	"log"
+	"path/filepath"
 	"time"
+
+	client "github.com/elastic/go-lumber/client/v2"
 )
 
-const SecondsForRecheck = 60
+func setup() (configs.Configuration, *registry.Registry) {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	c := configs.Configuration{}
+	c.GetConfiguration()
+
+	filehandler.CreateEmptyJsonFileIfNotExists(c.Registry)
+	r := &registry.Registry{}
+
+	return c, r
+}
 
 func main() {
-	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	config, r := setup()
+
+	recheckTime := config.GetRecheckDuration()
+
+	conn, err := client.SyncDial(config.Host + ":" + config.Port)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	for {
-		log.Print("Testing :)")
-		time.Sleep(time.Second * SecondsForRecheck)
+		// We need to re-read the content of the registry each time we wanna check if
+		// there are new files to process
+		registryFileContent := filehandler.GetFileContent(config.Registry)
+		registry.ConvertJSONToRegistry([]byte(registryFileContent), r)
+
+		// Globbing all paths according to the pattern
+		gzFilesPaths, err := filepath.Glob(config.Path)
+
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if len(gzFilesPaths) < 1 {
+			log.Print("There aren't gz files to process for " + config.Path)
+			time.Sleep(recheckTime)
+			continue
+		}
+
+		gzFilesToProcess := getGzNotProcessed(gzFilesPaths, *r)
+
+		if len(gzFilesToProcess) < 1 {
+			log.Print("There aren't gz files to process for " + config.Path)
+			time.Sleep(recheckTime)
+			continue
+		}
+
+		_, err = conn.Send(getBatchToSend(gzFilesToProcess))
+
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		registry.StoreFilesIntoRegistry(gzFilesToProcess, *r, config.Registry)
+
+		time.Sleep(recheckTime)
+	}
+
+}
+
+func getGzNotProcessed(gzFiles []string, r registry.Registry) []string {
+	gzFilesNotProcessed := []string{}
+	for _, gzFilePath := range gzFiles {
+		if !r.IsFileInRegistry(gzFilePath) {
+			gzFilesNotProcessed = append(gzFilesNotProcessed, gzFilePath)
+		}
+	}
+	return gzFilesNotProcessed
+}
+
+// Unzip the Gz in the same directory where they are
+func getBatchToSend(gzFiles []string) []interface{} {
+	batch := make([]interface{}, 1)
+	i := 0
+	for _, gzFilePath := range gzFiles {
```

Comment:
Shouldn't we keep the index from the for loop as `i`? Otherwise, the batch slice will keep only the last file found, won't it?

Reply:
The index in 0 is just to create the batch slice since the first element. This `func` won't be ever called in case that there aren't files to process (checked before in `getGzNotProcessed`) and this is how `elastic` did it to create the event that can be received by logstash.

Reply:
Okay, I get what you say now, sorry for the misunderstood. I'm going to check this. I need to change the size of the slice that contain the event and also increment the iterator. Thanks for the explanation.

Reply:
I've modified it using a slice without specifying the quantity of elements and just appending every new event inside.

Checked in logstash also that the batch of events is well received.